jmin-jlink/README.md

```diff
@@ -0,0 +1,155 @@
+<!--
+Copyright (c) 2020, 2020 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# jmin-jlink
+
+Includes:
+- **hellomodule.HelloWorld**, a simple app that contains classes, methods and fields - some of which should be removed during minimization;
+- **MinimimizeClassesPlugin**, a jlink plugin that minimizes an image using jarmin/jmin (jarmin/jmin-related code is commented out)
+- **MinimizeClassesPluginAgent**, a java agent that allows a custom implementation of a plugin to be used with jlink
+
+---
+
+## How to Run
+This command runs jlink with the custom plugin (jlink option `--minimize-classes`), enabled by the java agent, 
+on a simple app contained in `hellomodule`. The resulting image is generated in the `build` directory, and output 
+from the command is piped to `minimizeclasses.out`.
+
+As this code involves Java modularity, Java 11 or above is required.
+
+From the `jmin-jlink` directory, run the following:
+
+#### Compile the hellomodule classes
+```
+javac -d out --module-path out src/module-info.java src/hellomodule/*.java
+```
+
+#### Compile the custom jlink plugin
+```
+javac --add-exports jdk.jlink/jdk.tools.jlink.plugin=jmin.jlink --add-exports jdk.jlink/jdk.tools.jlink.internal=jmin.jlink plugin/jmin.jlink/module-info.java plugin/jmin.jlink/jmin/jlink/plugin/*.java
+```
+
+#### Compile the java agent
+```
+javac --add-exports jdk.jlink/jdk.tools.jlink.internal=plugin.agent --add-exports jdk.jlink/jdk.tools.jlink.plugin=plugin.agent agent/plugin.agent/module-info.java agent/plugin.agent/jmin/jlink/agent/*.java
+cd agent/plugin.agent
+jar cfm jminjlinkagent.jar MANIFEST.mf jmin/jlink/agent/*.class *.class
```

Comment:
The instructions as is didn't work for me. In particular:
* the manifest is `MANIFEST.MF` (all upper-case)
* they advise writing to the `out` directory which is not mentioned in `.gitignore`.

I did get it all to work with this `Makefile`:
```
JAVA_HOME   ?= $(HOME)/space/openj9/jdk11/build/linux-x86_64-normal-server-release/images/jdk

JAR         := $(JAVA_HOME)/bin/jar
JAVAC       := $(JAVA_HOME)/bin/javac
JLINK       := $(JAVA_HOME)/bin/jlink

OUT         := build

all : \
	$(OUT)/hellomodule.marker \
	$(OUT)/jminjlinkagent.jar \
	$(OUT)/plugin.marker \
	#

clean :
	rm -rf $(OUT)

$(OUT)/jminjlinkagent.jar : \
		agent/plugin.agent/MANIFEST.MF \
		agent/plugin.agent/module-info.java \
		$(wildcard agent/plugin.agent/jmin/jlink/agent/*.java)
	@mkdir -p $(OUT)/agent
	$(JAVAC) \
		-d $(OUT)/agent \
		--add-exports jdk.jlink/jdk.tools.jlink.internal=plugin.agent \
		--add-exports jdk.jlink/jdk.tools.jlink.plugin=plugin.agent \
		agent/plugin.agent/module-info.java \
		agent/plugin.agent/jmin/jlink/agent/*.java
	$(JAR) \
		--create \
		--file=$(OUT)/jminjlinkagent.jar \
		--manifest=agent/plugin.agent/MANIFEST.MF \
		-C $(OUT)/agent module-info.class \
		-C $(OUT)/agent jmin

$(OUT)/hellomodule.marker : \
		src/module-info.java \
		$(wildcard src/hellomodule/*.java)
	@mkdir -p $(OUT)/modules/hellomodule
	$(JAVAC) -d $(OUT)/modules/hellomodule \
		src/module-info.java \
		src/hellomodule/*.java
	touch $@

$(OUT)/plugin.marker : \
		plugin/jmin.jlink/module-info.java \
		$(wildcard plugin/jmin.jlink/jmin/jlink/plugin/*.java)
	@mkdir -p $(OUT)/plugin
	$(JAVAC) \
		-d $(OUT)/plugin \
		--add-exports jdk.jlink/jdk.tools.jlink.internal=jmin.jlink \
		--add-exports jdk.jlink/jdk.tools.jlink.plugin=jmin.jlink \
		plugin/jmin.jlink/module-info.java \
		plugin/jmin.jlink/jmin/jlink/plugin/*.java
	touch $@

test : all
	$(JLINK) \
		-J-javaagent:$(OUT)/jminjlinkagent.jar \
		-J--add-modules=jmin.jlink \
		-J--module-path=:$(OUT)/plugin \
		--minimize-classes \
		--module-path=$(OUT)/modules \
		--add-modules hellomodule \
		--output $(OUT)/min \
			2>&1 | tee minimizeclasses.out
```

Reply:
I'm not entirely sure why the instructions aren't working as is. Perhaps the paths to javac/jar/jlink were not all aligned? Running each of the commands with the absolute path to javac/jar/jlink it appears to work, so I will update the build instructions to include `<PATH_TO_JAVA_BIN>/[javac/jar/jlink]`...although not very ideal.

Do you recommend including a Makefile like this one or adding a bash script for convenience?

Reply:
Did you test this on Windows? Or a system with a case-sensitive filesystem?

My preference (in case you couldn't guess) is for a Makefile, which allows for expressing when the steps need to be executed (and when they can be skipped). You'll note I avoided mingling the `.class` files among the source files.

If you add a Makefile, it should demand the user specify `JAVA_HOME` (possibly in the environment).

Reply:
I've updated the README and tested the manual instructions on Mac and Windows -- working for both.

Added a Makefile which works on Mac. Not sure if I should extend the Makefile to be portable to Windows? There's likely some incompatibilities with the path separator and commands such as `rm` on Windows.

Reply:
I think it's reasonable to assume cygwin tools are available in which case the Makefile as written should work even on Windows.